#' Differential methylation evaluation over time for scDNA-seq data under the 2-group scenario.
#'
#' This function performs DM analysis to identify genomic features showing drastic
#' changes between two groups along pseudotime, by modeling the methylation changes
#' and comparing the fitted curves for each group.
#'
#' @param beta_sigma_list_group A list containing two elements:
#' - `Group1`: A numeric list of estimated parameters for all genomic features in group 1 (generated by \code{estiParamTwoGroups}).
#' - `Group2`: A numeric list of estimated parameters for all genomic features in group 2.
#'
#' @return A named numeric vector where each value corresponds to a genomic feature
#' (e.g., a gene). The values represent the integral of the differences between the
#' fitted curves of scDNA methylation levels for the two groups. The values are sorted
#' in descending order, with larger values indicating more drastic differences between the groups.
#'
#' @import stats MCMCpack BiocParallel car mvtnorm
#' @importFrom S4Vectors subjectHits queryHits Rle
#' @importFrom methods is
#' @importFrom rtracklayer start offset end
#' @importFrom Matrix cov2cor toeplitz update
#' @export
dmTwoGroups <- function(beta_sigma_list_group) {
  # Check if input is valid
  if (is.null(beta_sigma_list_group$Group1) || is.null(beta_sigma_list_group$Group2)) {
    stop("Missing parameter lists for one or both groups!",
         call. = TRUE, domain = NULL
    )
  }

  # Extract the parameter lists for Group 1 and Group 2
  beta_mu_mean_group1 <- beta_sigma_list_group$Group1
  beta_mu_mean_group2 <- beta_sigma_list_group$Group2

  # Remove features with invalid values (NA/Inf) in either group
  bad_elements_group1 <- lapply(beta_mu_mean_group1, contains_inf_or_na)
  bad_elements_group2 <- lapply(beta_mu_mean_group2, contains_inf_or_na)

  # Keep only features that are valid in both groups
  valid_features <- !unlist(bad_elements_group1) & !unlist(bad_elements_group2)
  beta_mu_mean_group1 <- beta_mu_mean_group1[valid_features]
  beta_mu_mean_group2 <- beta_mu_mean_group2[valid_features]

  # Calculate the integral of differences between the two groups for each feature
  int_list <- bplapply(seq_along(beta_mu_mean_group1), function(i) {
    calculate_integral(beta_mu_mean_group1[[i]], beta_mu_mean_group2[[i]])
  }, BPPARAM = MulticoreParam())

  # Assign names of the valid features to the result
  names(int_list) <- names(beta_mu_mean_group1)

  # Convert the list of integrals to a named numeric vector and sort in descending order
  int_res <- unlist(int_list)
  int_res <- sort(int_res, decreasing = TRUE)

  return(int_res)
}
